### PR TITLE
Pause / Resume Workflow activity

### PR DESCRIPTION
**Pull request by DeepanshuA**
# Description

Changes to pause/resume a workflow instance.

## Issue reference

Please reference the issue this PR will close: #6042 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: dapr/proto/runtime/v1/dapr.proto [hunk 1]
```diff
@@ -703,12 +709,12 @@ message StartWorkflowRequest {
...
 
-message TerminateWorkflowRequest {
+message WorkflowActivityRequest {
   string instance_id = 1;
   string workflow_component = 2;
 }
 
-message TerminateWorkflowResponse {
+message WorkflowActivityResponse {
 }
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply dapr/proto/runtime/v1/dapr.proto:718]: It’s a breaking change in the generated SDK names but not in the protobuf messages.
There are other, bigger breaking changes in this API anyways.

**Comment by RyanLettieri** [comment]
> can you add more unit tests to both universal api file and the component.go file in wfenginge?

+1 to this. A test or 2 in the wfengine_test.go for Pause and Resume would be useful.

File: dapr/proto/runtime/v1/dapr.proto (hunk 1 above)
[the remark this answers is not in the document]
Comment (DeepanshuA) [reply dapr/proto/runtime/v1/dapr.proto]: Yeah, this would be a breaking change.

**Comment by DeepanshuA** [comment]
> can you add more unit tests to both universal api file and the component.go file in wfenginge?

Added UTs

File: dapr/proto/runtime/v1/dapr.proto [hunk 2]
```diff
@@ -701,12 +707,12 @@ message StartWorkflowRequest {
...
-  string instance_id = 1 [json_name = "instanceId"];
-  string workflow_component = 2 [json_name = "workflowComponent"];
+message WorkflowActivityRequest {
+  string instance_id = 1;
+  string workflow_component = 2;
 }
 
-message TerminateWorkflowResponse {
+message WorkflowActivityResponse {
 }
```
Comment (ItalyPaleAle) [inline dapr/proto/runtime/v1/dapr.proto:716]: Nit, just so it's clear it was intentional :)

```suggestion
message WorkflowActivityResponse {
  // Empty
}
```

File: pkg/grpc/universalapi/api_workflow.go [hunk 3]
```diff
@@ -189,3 +189,69 @@ func (a *UniversalAPI) RaiseEventWorkflowAlpha1(ctx context.Context, in *runtime
 	}
 	return &runtimev1pb.RaiseEventWorkflowResponse{}, nil
 }
+
+func (a *UniversalAPI) PauseWorkflowAlpha1(ctx context.Context, in *runtimev1pb.WorkflowActivityRequest) (*runtimev1pb.WorkflowActivityResponse, error) {
```
Comment (ItalyPaleAle) [inline pkg/grpc/universalapi/api_workflow.go:193]: This, `TerminateWorkflowAlpha1`, and `ResumeWorkflowAlpha1` are pretty much identical, even the signature is the same. The only difference seems to be that one calls `Pause`, one calls `Terminate`, and one calls `Resume`.

Please consider merging them into a single shared function. Something like a private function that accepts `(ctx, in, method)` where `method` is like `workflowComponent.Pause`.

Less code = easier to maintain :)

File: pkg/grpc/universalapi/api_workflow_test.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.testName, func(t *testing.T) {
+			req := &runtimev1pb.WorkflowActivityRequest{
+				InstanceId:        tt.instanceId,
+				WorkflowComponent: tt.workflowComponent,
+			}
+			var err error
+			if tt.apiToBetested == PauseWorkflow {
```
Comment (ItalyPaleAle) [inline pkg/grpc/universalapi/api_workflow_test.go:142]: Can you test Terminate too? Since it's pretty much the same

File: pkg/http/api.go [hunk 5]
```diff
@@ -827,8 +839,8 @@ func (a *api) onGetWorkflowHandler() fasthttp.RequestHandler {
 func (a *api) onTerminateWorkflowHandler() fasthttp.RequestHandler {
```
Comment (ItalyPaleAle) [inline pkg/http/api.go:845]: Here too, `onTerminateWorfklowHandler()`, `onResumeWorkflowHandler()` and `onPauseWorkflowHandler()` are pretty much identical down to th signature of `InModifier`. Please consider merging them into a single method

Comment (ItalyPaleAle) [reply pkg/http/api.go:845]: Example:

```go
type workflowActivityHandler = func(ctx context.Context, in *runtimev1pb.WorkflowActivityRequest) (*runtimev1pb.WorkflowActivityResponse, error)

// Used by TerminateWorkflowAlpha1, PauseWorkflowAlpha1, ResumeWorkflowAlpha1
func (a *api) onWorkflowActivityHandler(handler workflowActivityHandler) fasthttp.RequestHandler {
	return UniversalFastHTTPHandler(
		handler,
		UniversalFastHTTPHandlerOpts[*runtimev1pb.WorkflowActivityRequest, *runtimev1pb.WorkflowActivityResponse]{
			InModifier: func(reqCtx *fasthttp.RequestCtx, in *runtimev1pb.WorkflowActivityRequest) (*runtimev1pb.WorkflowActivityRequest, error) {
				in.InstanceId = reqCtx.UserValue(instanceID).(string)
				in.WorkflowComponent = reqCtx.UserValue(workflowComponent).(string)
				return in, nil
			},
			SuccessStatusCode: fasthttp.StatusAccepted,
		})
}
```

Then just pass the handler in the router, for example:

```go
		{
			Methods: []string{fasthttp.MethodPost},
			Route:   "workflows/{workflowComponent}/{instanceID}/pause",
			Version: apiVersionV1alpha1,
			Handler: a.onWorkflowActivityHandler(a.universal.PauseWorkflowAlpha1),
		},
```

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: go.mod [hunk 6]
```diff
@@ -424,3 +424,5 @@ replace (
 //
 // Then, run `make modtidy` in this repository.
 // This ensures that go.mod and go.sum are up-to-date.
+
+replace github.com/dapr/components-contrib => github.com/DeepanshuA/components-contrib v1.6.0-rc.2.0.20230327134753-c1df8bfd5f5e
```
Comment (DeepanshuA) [inline go.mod:428]: Reviewers: I will remove it and refer contrib with interface changes, before this PR gets ready to be merged.

File: pkg/grpc/universalapi/api_workflow_test.go (hunk 4 above)
Comment (DeepanshuA) [reply pkg/grpc/universalapi/api_workflow_test.go:182]: Done

File: pkg/http/api.go (hunk 5 above)
Comment (DeepanshuA) [reply pkg/http/api.go]: Sure, done

**Review by ItalyPaleAle — APPROVED** [review]

**Comment by mukundansundar** [comment]
@DeepanshuA https://github.com/dapr/components-contrib/pull/2683 is merged, please change contrib

**Comment by DeepanshuA** [comment]
This was wrong on my part, I shouldn't have force-pushed. Now, the lint changes from Alessandro are gone. I am trying to search for the particular commit id to cherr-pick but am not able to find. Trying to search for it.
But, definitely a lesson for me, back to basics - shouldn't force push.

**Review by ItalyPaleAle — APPROVED** [review]

**Review by ItalyPaleAle — APPROVED** [review]

**Pull request merged**